### PR TITLE
[universal-redux] Upgrading to react-router 4.x

### LINKED ISSUE #161: Upgrade to React Router 4
image

It would be awesome if someone could help with this transition. Please refer to this comment: https://github.com/diegohaz/arc/issues/60#issuecomment-285832122

RR4 has [new docs](https://reacttraining.com/react-router/web/guides/quick-start) and redux integration is now part of the RR repo (https://reacttraining.com/react-router/web/guides/redux-integration)

Let's do that.

### PR DESCRIPTION
**Pull request by kybarg**
closes #161

### CODE REVIEW

**Comment by diegohaz** [comment]
Awesome, @kybarg 

Will review that asap

File: package.json [hunk 1]
```diff
@@ -49,11 +49,11 @@
     "nodemon": "^1.11.0",
     "opn-cli": "^3.1.0",
     "react-addons-test-utils": "^15.4.1",
+    "react-hot-loader": "^3.0.0-beta.6",
```
Comment (diegohaz) [inline package.json]: `react-hot-loader` needs to be a production dependency. The library itself handles `production`/`development` properly.

File: package.json [hunk 2]
```diff
@@ -65,13 +65,14 @@
...
-    "copyfiles": "^1.0.0",
+    "copyfiles": "^1.2.0",
     "cross-env": "^3.1.4",
     "csurf": "^1.9.0",
     "express": "^4.14.0",
     "express-force-ssl": "^0.3.2",
     "file-loader": "^0.10.0",
-    "history": "3.0.0",
+    "history": "^4.6.1",
+    "isomorphic-fetch": "^2.2.1",
```
Comment (diegohaz) [inline package.json]: Could you please make that change (replacing `axios` by `isomorphic-fetch`) on another PR? It's actually a bug we missed.

File: src-clean/client.js [hunk 3]
```diff
@@ -3,25 +3,24 @@ import React from 'react'
...
 
-import routes from 'routes'
+import AppRoutes from 'routes'
 
 // eslint-disable-next-line no-underscore-dangle
 const initialState = window.__INITIAL_STATE__
-const baseHistory = useRouterHistory(createHistory)({ basename })
-const store = configureStore(initialState, baseHistory)
-const history = syncHistoryWithStore(baseHistory, store)
+const history = createHistory()
```
Comment (diegohaz) [inline src-clean/client.js]: What about `basename`?

File: src-clean/server.js [hunk 4]
```diff
@@ -4,63 +4,70 @@ import serialize from 'serialize-javascript'
...
-      components.forEach((component) => {
-        if (component) {
-          while (component && !component[method]) {
-            // eslint-disable-next-line no-param-reassign
-            component = component.WrappedComponent
-          }
-          component &&
-          component[method] &&
-          promises.push(component[method]({ req, res, params, location, store }))
+  const fetchData = () => new Promise((resolve, reject) => {
```
Comment (diegohaz) [inline src-clean/server.js]: How does `fetchData` work now? Is that still asynchronous?

Comment (protoEvangelion) [inline src-clean/server.js]: @diegohaz @kybarg I'm testing this out right now.  Everything looks A ok 👍 When I make a post, I get this error `Encountered two children with the same key, `1:$101`` 

**_Is that expected?_**

Comment (kybarg) [inline src-clean/server.js]: @protoEvangelion Can't reproduce on test environment. Could you please provide more details?

Comment (kybarg) [inline src-clean/server.js]: @diegohaz Sure, nothing changed in general. Only `react-router` match logic has changed.

File: package.json (hunk 2 above)
Comment (kybarg) [inline package.json]: @diegohaz done

**Comment by diegohaz** [comment]
Thank you, @kybarg ❤️

Will review that this weekend.

File: src-clean/server.js (hunk 4 above)
Comment (protoEvangelion) [inline src-clean/server.js]: @kybarg Sorry for the delay!  This was not an issue with the upgrade.  **The problem existed before the upgrade.**  It has to do with the front end adding identical keys when creating a new post.  I will address this after **React-Router-4** is merged :)

**Comment by Geczy** [comment]
How'd the review go @diegohaz

**Comment by kybarg** [comment]
@diegohaz Any updates or your are closing this branch in favor for `redux-ssr`?

**Comment by diegohaz** [comment]
I still didn't have time to proper review that. Will do as soon as possible.

**Comment by Geczy** [comment]
Can you delegate anyone else to review this? Seems like you don't have much time to, would prefer quicker updates

**Comment by diegohaz** [comment]
@Geczy It's open for anyone who wants to test it. The boilerplate itself doesn't need to be updated for people update their projects. Please, try @kybarg's branch and give your feedback.

**Comment by Geczy** [comment]
Thanks for keeping this branch up to date @kybarg

**Comment by kybarg** [comment]
@Geczy You are welcome 😄

**Comment by Geczy** [comment]
@kybarg can you merge the latest redux branch?

**Comment by Geczy** [comment]
Running `yarn storybook` I get this error on all the Page books:

```
Cannot read property 'route' of undefined
TypeError: Cannot read property 'route' of undefined
    at Route.computeMatch (webpack:///./~/react-router/Route.js?:68:22)
    at new Route (webpack:///./~/react-router/Route.js?:47:20)
    at eval (webpack:///./~/react-dom/lib/ReactCompositeComponent.js?:295:18)
    at measureLifeCyclePerf (webpack:///./~/react-dom/lib/ReactCompositeComponent.js?:75:12)
    at ReactCompositeComponentWrapper._constructComponentWithoutOwner (webpack:///./~/react-dom/lib/ReactCompositeComponent.js?:294:16)
    at ReactCompositeComponentWrapper._constructComponent (webpack:///./~/react-dom/lib/ReactCompositeComponent.js?:280:21)
    at ReactCompositeComponentWrapper.mountComponent (webpack:///./~/react-dom/lib/ReactCompositeComponent.js?:188:21)
    at Object.mountComponent (webpack:///./~/react-dom/lib/ReactReconciler.js?:46:35)
    at ReactCompositeComponentWrapper.performInitialMount (webpack:///./~/react-dom/lib/ReactCompositeComponent.js?:371:34)
    at ReactCompositeComponentWrapper.mountComponent (webpack:///./~/react-dom/lib/ReactCompositeComponent.js?:258:21)
```

**Comment by kybarg** [comment]
@Geczy `redux` or `redux-ssr`?

**Comment by Geczy** [comment]
`kybarg:univeral-redux-rr4`

**Comment by Geczy** [comment]
what's the word @kybarg

**Comment by protoEvangelion** [comment]
@kybarg @Geczy Shouldn't we merge this into the `redux-ssr` branch because `universal-redux` is phasing out?

**Comment by diegohaz** [comment]
I'm looking into this right now. Brace yourselves, merge is coming.

**Comment by diegohaz** [comment]
Sorry for the delay and thank you very much, @kybarg and everyone.

The current `redux-ssr` is experimental (it's actually broken right now). `universal-redux` is still the official branch. I just need to take a moment to get it up to date with `master`/`redux`. Any help is welcome.

**Pull request merged**